docs/upgrading.md

```diff
@@ -10,6 +10,29 @@ We're building on the version `v3.4.0` a forward compatibility layer to help use
 To help on the migration process, all deprecated components are being marked with `@deprecated` and deprecated behaviour will trigger a `E_USER_DEPRECATED` error.
 However, you can also find here the instructions on how to make your code compatible with both versions.
 
+### General migration strategy
+
+It is recommended to update your existing software from any previous 3.x version to the latest 3.4.x version using 
+composer:
+
+    composer require lcobucci/jwt ^3.4
+
+Then run your tests and fix all deprecation notices.
+
+Note that PHPUnit will only break tests if you have the `E_USER_DEPRECATED` error level activated (it is a good idea to
+run tests using `E_ALL`, btw.), and deprecation messages during the execution of data providers will not affect the 
+test result at all, only output text to the console, because they are called before test execution starts. Make sure 
+you do not see any of them before you continue.
+
+Now you can upgrade to the latest 4.x version:
+
+    composer require lcobucci/jwt ^4.0
+
+At this stage, you'll probably find some more rough edges. Not every method call in the 3.4 version is
+100% identical to the 4.0 counterpart. So feel free to optimize your code at this point to leverage the
+core improvements like the configuration object, validation being separated from the `Token` value object and such. 
+The details are listed below.
```

Comment:
Is there any rough edge that you've found apart from the cases documented in this file? If not, why this paragraph is relevant?

The idea of this doc section is to be a step-by-step kind of thing, so people would go through the things below and modify their code. I'm confused with the need for summarising the changes.

Reply:
Rephrased and put emphasis on deprecation messages, and using the new features detailed later in this document.